core/src/Kokkos_View.hpp

```diff
@@ -1695,14 +1699,50 @@ class View : public ViewTraits<DataType, Properties...> {
     return map_type::memory_span(arg_layout) + scratch_value_alignment;
   }
 
+  template <typename PointerType>
+  explicit KOKKOS_INLINE_FUNCTION View(
+      const Kokkos::ScratchMemorySpaceBase<typename traits::execution_space,
+                                           PointerType>& arg_space,
+      const typename traits::array_layout& arg_layout)
+      : View(Impl::ViewCtorProp<pointer_type>(
+                 static_cast<pointer_type>(arg_space.template get_shmem_aligned(
+                     map_type::memory_span(arg_layout),
+                     scratch_value_alignment))),
+             arg_layout) {}
+
   explicit KOKKOS_INLINE_FUNCTION View(
       const typename traits::execution_space::scratch_memory_space& arg_space,
       const typename traits::array_layout& arg_layout)
-      : View(Impl::ViewCtorProp<pointer_type>(reinterpret_cast<pointer_type>(
+      : View(Impl::ViewCtorProp<pointer_type>(static_cast<pointer_type>(
```

Comment:
The static_cast here and below is a good point but I almost feel like it should be proposed as its own PR